includes/class-wc-payment-gateway-wcpay.php

```diff
@@ -1375,28 +1375,52 @@ public function process_payment_for_order( $cart, $payment_information, $schedul
 	 */
 	public function get_payment_method_to_use_for_intent() {
 		if ( WC_Payments_Features::is_upe_deferred_intent_enabled() ) {
-			return $this->get_payment_methods_from_request()[0];
+			$request_payment_method = sanitize_text_field( wp_unslash( $_POST['payment_method'] ?? '' ) ); // phpcs:ignore WordPress.Security.NonceVerification
+			return $this->get_payment_methods_from_gateway_id( $request_payment_method )[0];
 		}
 	}
 
 	/**
-	 * Get the payment methods used in the request.
+	 * Get payment method types to attach to intention request.
 	 *
+	 * @param Payment_Information $payment_information Payment information object for transaction.
 	 * @return array List of payment methods.
 	 */
-	private function get_payment_methods_from_request() {
-		$upe_payment_method = sanitize_text_field( wp_unslash( $_POST['payment_method'] ?? '' ) ); // phpcs:ignore WordPress.Security.NonceVerification
+	public function get_payment_method_types( $payment_information ) {
+		$request_payment_method = sanitize_text_field( wp_unslash( $_POST['payment_method'] ?? '' ) ); // phpcs:ignore WordPress.Security.NonceVerification
+		$token                  = $payment_information->get_payment_token();
+
+		if ( ! empty( $request_payment_method ) ) {
+			$payment_methods = $this->get_payment_methods_from_gateway_id( $request_payment_method );
+		} elseif ( ! is_null( $token ) ) {
+			$order           = $payment_information->get_order();
+			$order_id        = is_a( $order, 'WC_Order' ) ? $order->get_id() : null;
```

Comment:
Let's use `$order instanceof WC_Order` - a better practice.

Reply:
Good call--updated in 178d0760d108485e4ffa4debf73c0fa62ad35969.